src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -202,19 +306,370 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+                        IReadOnlyList<Descriptor> relatedDescriptors = this.bulkUpdateGraph.GetRelatedDescriptors(lastVisitedDescriptor);
+
+                        if (relatedDescriptors.Count == 0)
+                        {
+                            // Check the index of the last visited descriptor in the descriptors list.
+                            // If the index is the same value as the length of the descriptors list then
+                            // all the descriptors in the descriptors list have been traversed and we pop the items from the stack. 
+                            if (adjacentDescriptors.Peek().LastVisited == adjacentDescriptors.Peek().Descriptors.Count - 1)
+                            {
+                                adjacentDescriptors.Pop();
+                            }
+
+                            int index = adjacentDescriptors.Peek().LastVisited;
+                            IReadOnlyList<Descriptor> descriptors = adjacentDescriptors.Peek().Descriptors;
+
+                            if (index == descriptors.Count - 1)
+                            {
+                                adjacentDescriptors.Pop();
+                            }
+                            else
+                            {
+                                // To get the next descriptor to be visited, we take the last visited index and increment it by one.
+                                // Then we get the descriptor at the new index from the descriptors list. 
+                                // We then update the index to the new index value. 
+                                descriptor = descriptors[index + 1];
+                                adjacentDescriptors.Pop();
+                                adjacentDescriptors.Push((index + 1, descriptors));
+                            }
+                        }
+                        else 
+                        {
+                            // If we find the last visited descriptors had related descriptors then we start traversing the related descriptors list and
+                            // we start with the descriptor at index 0 and keep updating the index as we traverse the list. 
+                            descriptor = relatedDescriptors[0];
+                            adjacentDescriptors.Push((0, relatedDescriptors));
+                        }
+                    }          
+                }
+                 
+                if (descriptor != null) 
+                {
+                    this.SaveResultProcessed(descriptor, HasFailedOperation(materializerState));
+
+                    operationResponse = new ChangeOperationResponse(this.cachedResponse.Headers, descriptor)
+                    {
+                        StatusCode = (int)this.cachedResponse.StatusCode
+                    };
+
+                    this.materializerStateForDescriptor.Add(descriptor, materializerState);
+
+                    if (this.cachedResponse.Exception != null)

```

Comment:
Is there a test where you mock a response from the service that gives rise to this scenario?

Reply:
https://github.com/OData/odata.net/blob/35b1f192c7b1bc8f80d2d0456ebbe8c9290602ef/test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs#L169

Have moved this check to this method `HandleResponse`